### ISSUE TITLE
Support for changesets in WordPress 4.7?

### ISSUE THREAD

**Issue by pavelevap** [versionpress/versionpress#1166]
Users can make changes in Customizer (change theme, styling, etc), copy URL of these changes with parameter `changeset_uuid=ee56fbfc-c7bb-43f7-870e-e4beb0e45b05` (without saving changes) and share it with somebody for other updates and discussion. Special post type `customize_changeset` with postname `ee56fbfc-c7bb-43f7-870e-e4beb0e45b05` is created as `auto-draft` (**VP does not track** `post_status` set to `auto-draft` currently).

But after publishing these changes, particular post is trashed (later probably deleted by cron as usual - I did not check it yet) and no longer available for use (but **VP is tracking** this change). Changesets are probably not used for revision purposes currently, but it could change in the future.

So, in the end, **VP is tracking what is not needed** (trashed changesets) **and ignoring what is needed** :-)

https://make.wordpress.org/core/2016/10/12/customize-changesets-formerly-transactions-merge-proposal/

There are two possibilities:

1) Changesets can be ignored, but then VP will stop tracking this WP 4.7 feature. We can add `post_type: customize_changeset` to `ignored-entities` and it is done. But I can imagine, that somebody can really use this special `changeset_uuid` parameter to share proposed theme changes (it can be a lot of work to setup all options for some themes) and in the end it can be lost (and also not available to share between environments).

2) VP can start supporting this feature, but there are some things to do:

- Start tracking `auto-draft` for particular `post_type: customize_changeset` (probably not supported by VP currently).

- Track entities in saved data (see attached screenshot). Saved data are stored in some kind of JSON format in `post_content`, so we should be able to track something like `customize_changeset - post_content - option - user_id: user` (generally, many rules from `schema.yml` will be applied here). Related: #1145

- And it also does not look very good in administration.

I am not sure about it, any thoughts? In the end, we should probably support it, I guess.

![image](https://cloud.githubusercontent.com/assets/3008007/21058393/0a67d736-be3e-11e6-9400-75ca06d6a5d4.png)

**Comment by pavelevap** [comment]
Additional info: `customize_changeset` is available only for 7 days and then deleted as a part of regular garbage collection script which handles all `auto-drafts`. It is a little bit strange, because users can loose their changes without notification, but it works this way currently.

**Comment by borekb** [comment]
Thanks Pavel. I'm a big fan of Customizer changesets but the way I see them is that they are temporary states of a site, meant to live relatively shortly – they are not true states worth recording. Maybe we can get a confirmation from @westonruter?

It seems to me that we should just fix the bug which tracks changes to the trashed changeset.

On another note, I have a feeling that changesets are going to be part of the [big transformation](https://blog.versionpress.net/2016/12/state-of-the-word-2016-the-significant-one/) that WordPress is undergoing in 2017. In a sense, they are a staging concept which is also a core concept of VersionPress so we should watch this area closely and align our efforts.

**Comment by westonruter** [comment]
@borekb hey, thanks for looping me in.

The `customize_changeset` posts are stored with `auto-draft` status by default so that they will get garbage-collected after a week, just like for the posts that are creared when landing on `post-new.php` if the post is never saved as a draft. When the `customize_changeset` post transitions to the `publish` status (that is, when the user presses “Save & Publish) the values for the settings then get saved into their various locations in the DB according to their respective `WP_Cutomize_Setting::update()` implementations, and the `customize_changeset` post then by default will then automatically get transitioned to `trash` (which then get deleted after 30 days, by default). Since there is no UI in core (yet) for changesets, it seemed important to garbage-collect all of these posts. However, if you look at the [Customize Changesets Technical Design Decisions](https://make.wordpress.org/core/2016/10/12/customize-changesets-technical-design-decisions/) post, you'll read this:

> By adding adding support for `revisions` to the `customize_changeset` post type, published changesets will no longer be automatically trashed (and garbage-collected). With this there is automatic revisions and an audit trail for changes made in the customizer (see [#31089](https://core.trac.wordpress.org/ticket/31089)).

In fact, this will soon be implemented by the [Customize Snapshots](https://github.com/xwp/wp-customize-snapshots) feature plugin which will serve as the plugin for providing the UI for changesets until it gets into core, or some iteration on the UI.

With `revisions` support enabled, changeset posts then do serve as a revision history for all changes made to your site (as long as you make those changes via the customizer). You can walk back over previous published `customize_changeset` posts to see how the setting values change over time. It would also be possible just before publishing a changeset to capture the previous values for the settings to store with that changeset so that a revert changeset could later be computed (see https://github.com/xwp/wp-customize-snapshots/issues/20).

So yeah, in 4.7 the `customize_changeset` posts are normally going to be short-lived. A user would only be able to keep a changeset alive by bookmarking the URL that includes the `changeset_uuid` to continue making modifications, and continuing to keep pushing-out the 7-day expiration time. In the future (and with the Customize Snapshots feature plugin) these changeset posts can be long-lived drafts and even scheduled to be applied in the future via the `future` post status.

Naturally with long-lived changesets there are opportunities for conflicts between concurrently-edited changesets. These have been raised as issues in the Customize Snapshots plugin, but we haven't given it anywhere near the focus that you have with VP.

**Comment by pavelevap** [comment]
After this discussion I would still vote to support `customize_changeset` (even if much harder than ignoring it):

- Current temporary state does not mean that users could loose huge amount of work. This `auto-draft` is **different** from standard `post`, because it can contain very important data for users (and is used for other work and possibly cooperation).
- Expiration time of 7 days can be extended for `customize_changeset` (not possible for `post`, I guess) simply by making any change (date for garbage collection changed).
- Example use case: Start website with Twenty Sixteen, buy premium theme, prepare changes in Customizer (during several days and cooperation with colleague with the help of `customize_uuid`) and publish changes in the end after review (themes can have many different settings, widget areas, etc). This is nice feature which can be pretty common after some more tutorials and plugins appear and users should not loose their work.
- Release date of WordPress 4.8 is not known and it can be much longer period to wait (and ignore this feature). Also we are not sure that there will be any proposed changes implemented in next version.
- Supporting this feature we can help to save work of some users which are not familiar with important detail that changes are saved only for 7 days :-)

**Comment by westonruter** [comment]
Personally I think that VP should ignore `customize_changeset` posts that have the `auto-draft` status.

**Comment by borekb** [comment]
@westonruter Great insights, thanks!

**Comment by borekb** [comment]
@pavelevap You're basically proposing taking on some responsibilities that changesets intentionally avoided. I agree that some aspects of changesets are a bit problematic at this point (it's a v1 release after all) but I don't think VersionPress should be a tool for users to work around them.

I'm more interested in watching where changesets go in the future and embracing them in a broader sense. For example, if they have challenges with merging and conflict resolution, we could provide the necessary infrastructure for that. If they and up providing some sort of staging UI, we could embrace it for our (more complete) staging. Things like that.

In the long term, I think that changesets and VersionPress align pretty well.

**Comment by pavelevap** [comment]
@borekb: Nope :-) I only want to support this feature by default because it can contain important data for users, nothing more. Currently it is temporary by design, but it can be temporary for many weeks and contain a huge amount of work. The main problem is that we are not able to distinguish important data for user and common Customizer garbage because everything is `auto-draft`. But as a user, I would like to have all important data available during potential restore (and especially rely on VP) and it would be also helpful to transfer it between staging and production (even if it can be replaced by merging `theme_mod_*` value).

By implementing it can VP take one more step forward (enable to find automatically deleted data in history), but it is only default VP feature :-)

**Comment by borekb** [comment]
I don't agree. Changesets intentionally decided to use `auto-draft`, auto-drafts are garbage-collected by nature and VersionPress does not track such things. We would need to have a really strong case to add specific workarounds for changesets, for example, if examples such as yours were repeated over and over again. Until that happens, I think we have other areas to invest our time in.

**Comment by pavelevap** [comment]
No problem :-) Still, I will try to explain my point in detail for possible future implementation:

`auto-draft (post) != auto-draft (customize_changeset)`

`auto-draft` was generally designed to avoid race conditions during creation of new post and that is why it contains `auto-` part. It contains not important data (created `auto-`matically by WordPress) which are not useful **after several seconds** and VP should not track them. `auto-draft` is changed into `draft` status almost immediatelly (and `draft` is tracked by VP). `draft` is something edited (and editable) by user and stored in database until published or deleted.

Let me [quote](http://stackoverflow.com/questions/10234271/wordpress-auto-draft-disabling) others: 

>Auto-drafts exist because of the fact that multiple users can create new posts at the same time. If two people enter post-new at roughly the same moment, then have their first autosaves occur nearly simultaneously, then there is a race condition that can cause one of them to get back the wrong post ID, which will cause a post to be overwritten/lost when they then continue editing the post.
>
> The auto-draft creates the post and gets the ID of the new post before the editing screen is displayed, thus preventing two simultaneous authors from accidentally having the same post ID in the data in their browser.

So, no need to worry about `auto-draft` for posts and pages, because it is only "technical" status and lives several seconds (and deleted after a week).

In the case of `customize_changeset` VP should consider it more like `draft` (from data point of view) even if WordPress uses default `auto-draft` status. I believe it was a little bit unfortunate design decision because it is automatic only in the beginning (when user visits Customizer) and then everything is created by users manually (even if there is no UI yet). WordPress is using `auto-draft` for `customize_changeset` because it was probably easy solution (no buttons for saving), but with some negative side-effects. Garbage can be automatically collected, but user-created (and important) content can be also considered as garbage.

But even with these consequences, changesets can be used by users very well in WordPress 4.7, they can create them, edit them, share them and later publish them. I do not see any reason to ignore it. In this case we should be also ignoring `draft` because:

`draft (post) == auto-draft (customize_changeset)`

VP should consider importancy of data and not condemn them strictly based only on `auto-draft` status. `auto-draft` in this case can live for several weeks (not only seconds), it is editable by different users and it is part of working feature which can contain important data.

**Comment by westonruter** [comment]
The [Customize Snapshots](https://github.com/xwp/wp-customize-snapshots) plugin is all about providing the UI for transitioning a `customize_changeset` post from the `auto-draft` status to `draft` or another status. This UI for transitioning to `draft` I expect will be part of the next release of WP, whenever that is. Changesets that are `auto-draft` are not semantically intended to be different from Posts that have this same `auto-draft` status. We should really only be caring about `customize_changeset` posts that have transitioned from this initial transient status.

**Comment by pavelevap** [comment]
There were several changes during last period, see for example: https://core.trac.wordpress.org/changeset/40676 and https://core.trac.wordpress.org/changeset/40099

These changes suggested that `auto-draft` for `customize_changeset` is really **different** than `auto-draft` for standard `post` and VersionPress should track them (it is not a garbage, but important data for users which can be lost for example during cloning website).

No other changes for 4.8, I guess, so we will have to wait for changesets status and revisions for next (?) release.

**Comment by borekb** [comment]
Thanks for the links @pavelevap. Any fresh opinions from you, @westonruter?

**Comment by westonruter** [comment]
I don't think that `customize_changeset` posts should be tracked until they they transition from `auto-draft` status to another status at which point they no longer become transitory. That being said, when you add a new post or page in the customizer, a `post` is created also with the `auto-draft` status and it serves somewhat as a placeholder and stub for when the user will subsequently then flesh out in the admin presently, and in the future posts should be editable in the customizer itself with the changes being amended to the changeset. Presently when you transition a `customize_changeset` post from `auto-draft` to `draft` any such stub posts will not change their underlying status, but rather they will get their `post_date` bumped far in into the future to prevent automatic garbage collection after a week. So I can see that VersionPress _should_ actually start tracking any such `auto-draft` posts that were created as stubs for a `customize_changeset` once the changeset itself is no longer an `auto-draft`. Not sure how feasible that is, or if it would be just better to track all `auto-draft` posts regardless.

To play around with changeset status and revisions, you can install the [Customize Snapshots](https://github.com/xwp/wp-customize-snapshots) feature plugin.

**Comment by borekb** [comment]
Great info, thanks!
